Todo due date spec (#25)
ember-template-lint/ember-template-lint-todo-utils: docs/todo-due-dates.md

> - Feature Name: `due_dates`

Reviewer: This should probably match the filename.

- Start Date: `2020-11-09`
- RFC PR: (leave this empty)
- Checkup Issue: (leave this empty)

# Summary

[summary]: #summary

This RFC proposes enhancing the todo functionality with due days, allowing authors to add accountability and commitment to fix the todo items by a specified amount of time.

# Motivation

[motivation]: #motivation

> The lint todo functionality is intended to seamlessly integrate with existing linting tools, allowing the incremental fixing of linting errors on very large projects. However, one undesirable side effect of this is that developers may forget about fixing the todo items leading to the accumulation of technical debts.

Reviewer: ```suggestion
The lint todo functionality is intended to seamlessly integrate with existing linting tools, allowing the incremental fixing of linting errors on very large projects. However, one undesirable side effect of this is that developers may forget about fixing the todo items leading to the accumulation of technical debt.
```

> By adding a due date to the todo items, developers will now have a commitment to fix the linting problems because todo items that are past-due will once again be considered errors.

Reviewer: ```suggestion
By adding a time threshold to the todo items, developers will now have a commitment to fix the linting problems because todo items that are past this threshold will have their severity increased accordingly.
```

# Pedagogy

[pedagogy]: #pedagogy

- There will be documentation on how to configure when todo items are due to be fixed.

> - There will be a common config and CLI argument to specify the number of days todo items are due to be fixed.
> - There will be a common config and CLI argument to specify the number of days an item becomes an error after its due date.

Reviewer: ```suggestion
- There will be a config and/or CLI argument to specify the number of days an item's severity changes, either to `warn` or to `error`.
```

# Details

[details]: #details

## Requirements

- By default, todos are added with no due dates.
- Authors can specify the number of days the todos must be fixed after its creation date.

> - Authors can specify the number of days the todos are flagged as `error`s after they are past-due. The default value is `0`.

Reviewer: Why don't we instead allow for the configuration of both a warning threshold and an error threshold, individually. This would be the logic:

- if no warn or error thresholds are given, no due dates are configured
- if a warn threshold is given, in days, a todo transitions to a warn following that number of days past its creation date
- if an error threshold is given, in days, a todo transitions to an error following that number of days past its creation date
- if both a warn threshold _and_ and error threshold is given, a todo transitions first to a warn following that number of days past its creation date, then to an error following that number of days past the warn (warn + error days past creation date)

Author: I have two concerns with this:

- error threshold will have a different behavior if a warn threshold is also passed which is confusing. I believe it is better if an argument always has the same outcome. For example, error threshold is always the amount of time past its creation date.
- todos come from errors not warnings, so todos should not be in a state where they are marked as warnings permanently.

This is why I proposed `daysToFix`, where by default it is when todos become errors. If authors want a transitional period they can increase the additional number of days until it becomes an error.

Thoughts?

Reviewer: > error threshold will have a different behavior if a warn threshold is also passed which is confusing. I believe it is better if an argument always has the same outcome. For example, error threshold is always the amount of time past its creation date.

That's fine. I can understand this. We can make errors be an absolute value from creation date too, but it would need to be a number greater than warn if both were given.

> todos come from errors not warnings, so todos should not be in a state where they are marked as warnings permanently.

That's a valid point. What if we keep `daysToError` mandatory, but optionally allow `daysToWarn` to be provided, however you cannot use `daysToWarn` on its own. What I'm trying to do is keep the language consistent and inline with the severity terminology. Fix has a very specific meaning in linting, and it's not generally related to severity, which the todo feature _is_.

Author: Yes, I like that. I'll adjust.

- Todos that are past-due are flagged as `warning`s until the number of days to error is reached.
- Configuration can be specified in a common file or overriden via CLI arguments.

## Implementation

### High Level Flow

When an author runs a linter with the todo functionality, all errors are converted to todos with no due dates unless one is present in `package.json` or passed as a command line argument.

> - `daysToFix` is the number of days that the todo item needs to be fixed.
> - `daysToError` is the number of additional days past the `daysToFix` that the todo items are cosidered errors. Default is `0`.

Reviewer: I would suggest the following:

```suggestion
- `daysToWarn` - number of days after its creation date that a todo transitions into a warn
- `daysToError` number of days after its creation date that a todo transitions into an error _or_ if used with `daysToWarn`, the number of days after a todo transitions into a warn that it becomes an error
```

### Workflows

#### Normal todo workflow

- No due date is expected when adding todo items with no config.
- No due date is expected when adding todo items with no CLI param.

#### Todos that are past-due become errors

Use case: 20 days for todos to be considered errors since its creation date.

**Via config:**

In `package.json`:

 ```json
{
  "lint-todo": {

>     "daysToFix": 20,

Reviewer: ```suggestion
    "daysToError": 20,
```

  }
}
```

**Via `eslint` CLI:**

```

> UPDATE_TODO=1 DAYS_TO_FIX_TODO=20 yarn eslint . --format eslint-formatter-todo

Reviewer: ```suggestion
# to configure todos to transition to warnings 20 days after creation
UPDATE_TODO=1 DAYS_TO_WARN_TODO=20 yarn eslint . --format eslint-formatter-todo

# to configure todos to transition to errors 20 days after creation
UPDATE_TODO=1 DAYS_TO_ERROR_TODO=20 yarn eslint . --format eslint-formatter-todo
```

```

**Via `ember-template-lint` CLI:**

```

> yarn ember-template-lint . --update-todo --days-to-fix-todo 20

Reviewer: ```suggestion
# to configure todos to transition to warnings 20 days after creation
yarn ember-template-lint . --update-todo --days-to-warn-todo 20

# to configure todos to transition to errors 20 days after creation
yarn ember-template-lint . --update-todo --days-to-error-todo 20
```

```

#### Todos that are past-due become warnings, then errors

Use case: give 20 days for todos to be considered `warning`s, then 5 additional days until they are considered `error`s.

**Via config:**

```json
{
  "lint-todo": {

>     "daysToFix": 20,
>     "daysToError": 5

Reviewer: ```suggestion
    "daysToWarn": 20,
    "daysToError": 5
```

  }
}
```

**Via `eslint` CLI:**

```

> UPDATE_TODO=1 DAYS_TO_FIX_TODO=20 DAYS_TO_ERROR_TODO=5 yarn eslint . --format eslint-formatter-todo

Reviewer: ```suggestion
UPDATE_TODO=1 DAYS_TO_WARN_TODO=20 DAYS_TO_ERROR_TODO=5 yarn eslint . --format eslint-formatter-todo
```

```

**Via `ember-template-lint` CLI:**

```

> yarn ember-template-lint . --update-todo --days-to-fix-todo 20 --days-to-error-todo 5

Reviewer: ```suggestion
yarn ember-template-lint . --update-todo --days-to-warn-todo 20 --days-to-error-todo 5
```

```

### Proposed changes to the schema to include due date

```ts
interface TodoData {
  ...
  created: Date;

>   daysToFix?: int;
>   daysToError?: int;

Reviewer: ```suggestion
  daysToWarn?: int;
  daysToError?: int;
```

}
```

# Critique

[critique]: #critique

A few other options were considered:

- Instead of due days, another option considered was passing a specific due date. Passing days is preferable because if we want to expose a config/CLI param for this, passing a date would make the config or command to run stale.
- Allow the option to pass rule id with due date associated to them. The extra rule id config is not necessary since both `ember-template-lint` and `eslint` already have options to filter by rule id.
- If devs needs extra functionality, we could provide a different npx tool to do that. For now, the goal is to keep the todo functionality simple with a simple due date mechanism.

# Unresolved questions

> [unresolved]: #unresolved-questions

Reviewer: How will we ensure that if todos are completely regenerated (we run `UPDATE_TODO=1 or `--update-todo`) we don't lose this data?

Reviewer: I guess I'm not 100% sure if we should update the dates. If an initial due date is set, I think that should stand going forward. What do you think?

So, I think the second would be better. What circumstance would you want to refresh the todo due date?

Author: I can see the situation of people wanting to push the due date further because they didn't have enough time to fix the errors within X days, so they want to refresh the dates of all todos. WDYT?

---

Here's the proposal for the implementation:

When calling `writeTodos()` authors can now optionally pass `daysToError` and `daysToWarn`:

```ts
export declare function writeTodos(baseDir: string, lintResults: LintResult[], filePath?: string, daysToError?: int, daysToWarn?: int): Promise<string>;
```

**writeTodos() expected implementation:**

Writes files for todo lint violations. One file is generated for each violation, using a generated hash to identify each.

Given a list of todo lint violations, this function will also:

  - Delete existing files that no longer have a todo lint violation.
  - Keep existing files untouched for existing violations so as to not delete/update due days.

Adds the following for each todo item:

```ts
interface TodoData {
  ...
  daysToWarn?: int;
  daysToError?: int;
}
```

Note: `daysToWarn` and `daysToError` should not be considered in the hash generation for the violation.

**Linter expected implementation should be:**

- If `--update-todo` is passed to the Linter with `daysToWarn` and `daysToError`, `writeTodo()` is called and both arguments are passed. Todos will now have a due date associated to them.
- If error is present in the todo dir, check if `daysToWarn` threshold is reached, report as `warn`, if `daysToError` threshold is reached report as `error`, otherwise report as todo.

Reviewer: This seems good. 👍

- Feature Name: `todo-due-dates`
- Start Date: `2020-11-09`
- RFC PR: (leave this empty)
- Checkup Issue: (leave this empty)

# Summary

[summary]: #summary

This RFC proposes enhancing the todo functionality with due days, allowing authors to add accountability and commitment to fix the todo items by a specified amount of time.

# Motivation

[motivation]: #motivation

> The lint todo functionality is intended to seamlessly integrate with existing linting tools, allowing the incremental fixing of linting errors on very large projects. However, one undesirable side effect of this is that developers may forget about fixing the todo items leading to the accumulation of technical debt.

Reviewer: ```suggestion
The lint todo functionality is intended to seamlessly integrate with existing linting tools, allowing the incremental fixing of linting errors on very large projects. However, one undesirable side effect of this is that developers may forget about fixing the todo items, leading to the accumulation of technical debt.
```

> By adding a time threshold to the todo items, developers will now have a commitment to fix the linting problems because todo items that are past this threshold will have their severity increased accordingly.

Reviewer: ```suggestion
By adding a day threshold to the todo items, developers will now have a commitment to fix the linting problems because todo items that are past this threshold will have their severity increased accordingly.
```

> # Pedagogy

Reviewer 2: ```suggestion
# How We Will Teach This
```

suggesting this phrase because it's more consistent with the rest of the ecosystem

[pedagogy]: #pedagogy

- There will be documentation on how to configure when todo items are due to be fixed.
- There will be a config and/or CLI argument to specify the number of days an item's severity changes, either to `warn` or to `error`.

# Requirements

[requirements]: #requirements

> - By default, todos are added with no due dates.

Reviewer 3: Please define "by default" (aka be more specific). I _think_ you mean when no config is provided.

- Authors can specify the number of days the todos become errors after its creation date.

> - Author can also specify the number of days the todos become warnings after its creation date. The error threshold must also be passed and must be greater than this threshold.

Reviewer 3: I don't think we should require both warn + error together. I think it should be perfectly valid to allow folks to set a `daysToWarn` and not a `daysToError` setting. We should give guidance on what we think is best practice (setting both warning and error).

For example, I think it would be a perfectly valid setup to set `daysToWarn: 0` without ever transitioning to errors. In that world view, you might want to see them as warnings (as opposed to completely silencing/ignoring/squelching)  so that developers know that something is amiss when reviewing the file itself.

- Configuration can be specified in a common file or overriden via CLI arguments.

# Implementation

[requirements]: #implementation

## High Level Flow

> When an author runs a linter with the todo functionality, all errors are converted to todos with no due dates unless one is present in `package.json` or passed as a command line argument.

Reviewer: ```suggestion
When an author runs a linter with the todo functionality, all errors are converted to todos with no due date threshold unless one is present in `package.json` _or_ passed as a command line argument.
```

Reviewer: Would we want to be able to specify a different due date for eslint vs. ember-template-lint? Does it make sense to have this threshold only configured in package.json?

- `daysToError` number of days after its creation date that a todo transitions into an `error`.
- `daysToWarn` - number of days after its creation date that a todo transitions into a `warn`. If used, `daysToError` must also be passed and be greater than `daysToWarn`.

## Workflows

### Normal todo workflow

> - No due date is expected when adding todo items with no config or the CLI arguments `daysToError` or `daysToWarn`.

Reviewer: ```suggestion
- No due date is expected when adding todo items with no threshold provided in the `package.json` or the CLI arguments `daysToError` or `daysToWarn`.
```

### Todos that are past-due become errors

> Use case: 20 days for todos to be considered errors since its creation date.

Reviewer: ```suggestion
Example use case: 20 days for todos to be considered errors since its creation date.
```

**Via config:**

In `package.json`:

 ```json
{

>   "lint-todo": {

Reviewer: I'm unsure if we should use hyphen case here, particularly because you then use camel case below. I think we should consider using camel case for both.

>   "lintTodo": {
>     "daysToError": 20,
>   }

Reviewer 3: ```suggestion
  "lintTodo": {
    "daysToDecay": {
      "warn": 30,
      "error": 90
    }
  }
```

Where you could have:

* warn on its own
* error on its own
* both warn and error, but error must be greater than warn

What do you think?

Reviewer: I like this. Will make this change.

}
```

**Via `eslint` CLI:**

```
# to configure todos to transition to errors 20 days after creation
UPDATE_TODO=1 DAYS_TO_ERROR_TODO=20 yarn eslint . --format eslint-formatter-todo
```

**Via `ember-template-lint` CLI:**

```
# to configure todos to transition to errors 20 days after creation

> yarn ember-template-lint . --update-todo --days-to-error-todo 20

Reviewer 3: ```suggestion
yarn ember-template-lint . --update-todo --todo-days-to-error 20
```

I vaguely prefer this ordering, what do you think?

Reviewer: Ya I think that reads a bit better.

```

### Todos that are past-due become warnings, then errors

Use case: give 20 days for todos to be considered `warning`s, then 5 additional days until they are considered `error`s.

**Via config:**

```json
{

>   "lint-todo": {

Reviewer 3: ```suggestion
  "lintTodo": {
```

```

### Invalid option: only `daysToWarn` is present

If `daysToWarn` is passed without `daysToError` an error must be produced.

**Via `eslint` CLI:**

```
$ UPDATE_TODO=1 DAYS_TO_WARN_TODO=20 yarn eslint . --format eslint-formatter-todo

Error: `DAYS_TO_WARN_TODO` must be used with `DAYS_TO_ERROR_TODO`.
```

**Via `ember-template-lint` CLI:**

```
$ yarn ember-template-lint . --update-todo --days-to-warn-todo 20

Error: `--days-to-warn-todo` must be used with `--days-to-error-todo`.
```

#### Invalid option: `daysToWarn` is greater or equal than `daysToError`

If `daysToWarn` is greater or equal than `daysToError` an error must be produced:

**Via `eslint` CLI:**

```
$ UPDATE_TODO=1 DAYS_TO_WARN_TODO=20 DAYS_TO_ERROR_TODO=5 yarn eslint . --format eslint-formatter-todo

Error: `DAYS_TO_WARN_TODO` value must be less than `DAYS_TO_ERROR_TODO`.
```

**Via `ember-template-lint` CLI:**

```
$ yarn ember-template-lint . --update-todo --days-to-warn-todo 20 --days-to-error-todo=5

Error: `--days-to-warn-todo` value must be less than `--days-to-error-todo`.
```

## Proposed changes to the schema to include due date

```ts
interface TodoData {
  ...

>   created: Date;

Reviewer: I'd likely call this `createdDate`.

>   daysToWarn?: int;
>   daysToError?: int;

Reviewer 3: I feel like these should be actual dates, not offsets. Then this could become:

```suggestion
  warnDate?: Date;
  errorDate?: Date;
```

Reviewer: We initially wanted to store the offset so we wouldn't be serializing hard dates in the todos. That said, this is simply pre-calculating the date using the provided start date and offset, so I think this is fine. This also has the added benefit of being more readable.

}
```

# Critique

[critique]: #critique

A few other options were considered:

- Instead of due days, another option considered was passing a specific due date. Passing days is preferable because if we want to expose a config/CLI param for this, passing a date would make the config or command to run stale.
- Allow the option to pass rule id with due date associated to them. The extra rule id config is not necessary since both `ember-template-lint` and `eslint` already have options to filter by rule id.

> - If devs needs extra functionality, we could provide a different npx tool to do that. For now, the goal is to keep the todo functionality simple with a simple due date mechanism.

Reviewer: `npx` doesn't really have anything to do with this. It's rather that we could provide another CLI tool. `npx` is just a way we'd invoke it without installing the package.

# Unresolved questions

[unresolved]: #unresolved-questions

> - How will we ensure that if todos are completely regenerated (we run UPDATE_TODO=1 or --update-todo`) we don't lose this data?

Reviewer 3: TBH, I don't think we can, and **I** think that is a feature not a bug 😛

Reviewer: Fair.
